pkg/sql/schemachanger/sctest/comparator.go

```diff
@@ -163,32 +152,77 @@ func isInATransaction(ctx context.Context, t *testing.T, conn *gosql.Conn) bool
-// with different behavior under legacy vs under declarative schema changer.
-// Those cases are hard-coded, and if `line` is one of them, we transform it
-// into one with the same behavior under those two schema changers.
-func modifyBlacklistedStmt(
-	ctx context.Context, t *testing.T, line string, legacyConn *gosql.Conn,
-) string {
-	parsedLine, err := parser.Parse(line)
+// preExecutionProcessing is where we potentially modify input `line` so that
+// executing the returned line, one for LSC and one for DSC, results in the same
+// descriptor state in both clusters.
+func preExecutionProcessing(
+	ctx context.Context,
+	t *testing.T,
+	line string,
+	legacyConn *gosql.Conn,
+	declarativeConn *gosql.Conn,
+) (legacyLine string, declarativeLine string) {
+	if hasSyntaxError(line) {
+		return line, line
+	}
+
+	parsedLineForLegacy, err := parser.Parse(line)
 	require.NoError(t, err)
+	parsedLineForDeclarative, _ := parser.Parse(line)
 
+	// General, framework level modifications: They are applied regardless.
 	var modify bool
 	for _, lm := range []sqlLineModifier{
-		modifyExprsReferencingSequencesWithTrue,
-		modifyAlterPKWithRowIDCol,
 		modifySetDeclarativeSchemaChangerMode,
 		modifyCreateTempTable,
-		modifyAlterPKWithSamePKColsButDifferentSharding,
 	} {
 		var m bool
-		parsedLine, m = lm(ctx, t, parsedLine, legacyConn)
+		parsedLineForLegacy, m = lm(ctx, t, parsedLineForLegacy, legacyConn)
+		parsedLineForDeclarative, _ = lm(ctx, t, parsedLineForDeclarative, legacyConn)
 		modify = modify || m
 	}
+
+	// If `line` is supported in DSC, apply necessary modifications to ensure LSC
+	// and DSC converge after execution to account for known behavioral
+	// differences between the two schema changers.
+	declarativeLine = parsedLineForDeclarative.StringWithFlags(tree.FmtSimple | tree.FmtTagDollarQuotes)
+	if willLineBeExecutedInDSC(ctx, t, declarativeLine, declarativeConn) {
+		legacyLineModifiers := []sqlLineModifier{
+			modifyExprsReferencingSequencesWithTrue,
+			modifyAlterPKWithRowIDCol,
+			modifyAlterPKWithSamePKColsButDifferentSharding,
+		}
+		declarativeLineModifiers := []sqlLineModifier{
+			modifyExprsReferencingSequencesWithTrue,
+		}
+		for _, lm := range legacyLineModifiers {
+			var m bool
+			parsedLineForLegacy, m = lm(ctx, t, parsedLineForLegacy, legacyConn)
+			modify = modify || m
+		}
+		for _, lm := range declarativeLineModifiers {
+			parsedLineForDeclarative, _ = lm(ctx, t, parsedLineForDeclarative, legacyConn)
+		}
+	}
+
 	if modify {
-		t.Logf("Comparator testing framework modifies line %q to %q", line, parsedLine.String())
+		t.Logf("Comparator testing framework modifies line %q to %q", line, parsedLineForLegacy.String())
+	}
+
+	legacyLine = parsedLineForLegacy.StringWithFlags(tree.FmtSimple | tree.FmtTagDollarQuotes)
+	declarativeLine = parsedLineForDeclarative.StringWithFlags(tree.FmtSimple | tree.FmtTagDollarQuotes)
+	return legacyLine, declarativeLine
+}
+
+func willLineBeExecutedInDSC(
```

Comment:
I'm a bit confused about the need for this function. Couldn't the DSC modifiers perform a type switch to determine if a change is needed?